<a href="https://colab.research.google.com/github/rhythm38/woc5.0-ml-weather-prediction-case-study-RhythmArya/blob/main/WoC_Checkpoint_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# Importing necessary libraries


In [3]:
import numpy as np
import pandas as pd


#1) Reading the above CSV file in google colab using the pandas’ library and converting it into data frames.

In [5]:
df=pd.read_csv('/content/drive/MyDrive/WoC 5.0 ML/Dataset/Youtube Influencer Analysis - Updated.csv',encoding='cp1252')
df

Video Link Video Views  \
0          https://www.youtube.com/watch?v=_0fUOJ55gis     146,059   
1          https://www.youtube.com/watch?v=_70MSLPRfZI     231,969   
2    https://www.youtube.com/watch?v=_8gfCdt3H00&t=87s         878   
3          https://www.youtube.com/watch?v=_CkFNJgKsFs       9,965   
4          https://www.youtube.com/watch?v=_GFkHA5EZdE  39,780,561   
..                                                 ...         ...   
900        https://www.youtube.com/watch?v=GS18WawA-t0       6,892   
901        https://www.youtube.com/watch?v=_tc_APkjH2Q      16,554   
902        https://www.youtube.com/watch?v=Xfj4osSYfhU      32,798   
903        https://www.youtube.com/watch?v=mGEHqwVz0eI       1,503   
904        https://www.youtube.com/watch?v=UGATpt0I4vA     429,966   

                                           Video Title  \
0      Samse - Didupe Off-road Royal Enfield Himalayan   
1                             Bhoom Baddhal Web Series   
2    Customer Psychology, Four Views Of Consumer De...   
3       Top countries with proven natural gas reserves   
4                       73 Questions With Selena Gomez   
..                                                 ...   
900  Calming mashup of Islamic Nasheed 2022 | Isami...   
901                           Amchi Nawayat Mehfil hee   
902                                      Aas ge Kaagaz   
903  Anjuman Tarana by ABHS Students | Anjuman Anju...   
904  Jhulan Goswami Final Over in ODIs | ENG vs IND...   

                                           Channel URL          Creator Name  \
0                  https://www.youtube.com/c/sriharick            srihari ck   
1           https://www.youtube.com/c/ChaibisketTelugu           Chai Bisket   
2    https://www.youtube.com/channel/UCcIUnCsBfONvb...           Virda Warsi   
3                   https://www.youtube.com/c/trtworld                   TRT   
4                                      Vogue - YouTube                 Vogue   
..                                                 ...                   ...   
900        https://www.youtube.com/c/SahilOnlineTVnews  Sahil Online TV News   
901                https://www.youtube.com/sahilonline  Sahil Online TV News   
902                https://www.youtube.com/sahilonline  Sahil Online Bhatkal   
903                https://www.youtube.com/sahilonline  Sahil Online Bhatkal   
904        https://www.youtube.com/c/SonySportsNetwork   Sony Sports Network   

    Creator Gender Total Channel Subcribers Total Chanel Views  \
0             Male                    40200            7964284   
1             Male                   920000          137535681   
2           Female                      458              72945   
3              NaN                  1680000          541653219   
4              NaN                 12100000         3600496171   
..             ...                      ...                ...   
900            NaN                   594000          129926690   
901            NaN                   594000          129926690   
902            NaN                    23500            5104872   
903            NaN                    23500            5104872   
904            NaN                  3980000          728858923   

    Duration of Video Duration in Seconds  ... Video Description  Hashtags  \
0           00:30:41             1,841.00  ...               Yes         0   
1           00:14:24               864.00  ...               Yes         1   
2             0:08:12              492.00  ...               Yes         0   
3             0:01:06               66.00  ...               Yes         0   
4             0:07:40              460.00  ...               Yes         0   
..                ...                 ...  ...               ...       ...   
900          00:13:47              827.00  ...               Yes         0   
901          00:06:44              404.00  ...               Yes         2   
902          00:06:24              384.00  ...    

As we can see above csv file is read as dataframe only and dataframe contain 905 rows and 24 columns.

#2) How many values are missing for each column and try dropping or replacing NULL values using appropriate pandas’ function.

In [6]:
#below function shows no of missing values for each columns
df.isna().sum()

Video Link                                 0
Video Views                                0
Video Title                                0
Channel URL                                0
Creator Name                              11
Creator Gender                           314
Total Channel Subcribers                   0
Total Chanel Views                         0
Duration of Video                          0
Duration in Seconds                        0
Date of Video Upload                       0
No of Likes                                1
Language of the Video                     21
Subtitle                                   1
Video Description                          0
Hashtags                                   0
No of Comments                             0
Date of the Last Comment                  24
Maximum Quality of the Video               0
No of Videos the Channel                   0
No of Playlist                             9
Premiered or Not                           0
Community 

columns with missing values and no of missing values are as given below:
> Creator Name:11 <br>
> Creator Gender :314 <br>
> Language of the Video : 21 <br>
> Date of the Last Comment : 24 <br>
> Subtitle : 1 <br>
> No of Likes : 1 <br>


Dropping/replacing NULL values using appropriate pandas’ function.

In [7]:
df_with_missing_values_columns = df[['Creator Name', 'Creator Gender','Language of the Video','Date of the Last Comment','Subtitle','No of Likes']]
df_with_missing_values_columns

Creator Name Creator Gender Language of the Video  \
0              srihari ck           Male               Kannada   
1             Chai Bisket           Male                Telugu   
2             Virda Warsi         Female                 Hindi   
3                     TRT            NaN               English   
4                   Vogue            NaN               English   
..                    ...            ...                   ...   
900  Sahil Online TV News            NaN         Urdu + Arabic   
901  Sahil Online TV News            NaN             Nawayathi   
902  Sahil Online Bhatkal            NaN             Nawayathi   
903  Sahil Online Bhatkal            NaN                  Urdu   
904   Sony Sports Network            NaN               English   

    Date of the Last Comment Subtitle  No of Likes  
0                   8/6/2022       No       6700.0  
1                  6/15/2022       No       7800.0  
2                        NaN       No        858.0  
3                  9/22/2022       No        188.0  
4                  9/23/2022      yes     901000.0  
..                       ...      ...          ...  
900                10/4/2022       No        235.0  
901                10/4/2022       No        117.0  
902                 8/4/2022       No          1.0  
903                 8/4/2022       No          1.0  
904                10/4/2022      Yes       7600.0  

[905 rows x 6 columns]

In [8]:
df_with_missing_values_columns.dtypes

Creator Name                 object
Creator Gender               object
Language of the Video        object
Date of the Last Comment     object
Subtitle                     object
No of Likes                 float64
dtype: object

In [9]:
#example of dropping NULL values
df1 = df_with_missing_values_columns.dropna(subset=['Language of the Video'])
df1

Creator Name Creator Gender Language of the Video  \
0              srihari ck           Male               Kannada   
1             Chai Bisket           Male                Telugu   
2             Virda Warsi         Female                 Hindi   
3                     TRT            NaN               English   
4                   Vogue            NaN               English   
..                    ...            ...                   ...   
900  Sahil Online TV News            NaN         Urdu + Arabic   
901  Sahil Online TV News            NaN             Nawayathi   
902  Sahil Online Bhatkal            NaN             Nawayathi   
903  Sahil Online Bhatkal            NaN                  Urdu   
904   Sony Sports Network            NaN               English   

    Date of the Last Comment Subtitle  No of Likes  
0                   8/6/2022       No       6700.0  
1                  6/15/2022       No       7800.0  
2                        NaN       No        858.0  
3                  9/22/2022       No        188.0  
4                  9/23/2022      yes     901000.0  
..                       ...      ...          ...  
900                10/4/2022       No        235.0  
901                10/4/2022       No        117.0  
902                 8/4/2022       No          1.0  
903                 8/4/2022       No          1.0  
904                10/4/2022      Yes       7600.0  

[884 rows x 6 columns]

As we can see in above dataframe after dropping one row having NaN value,no of row reduced by one.

In [10]:
df_with_missing_values_columns["Creator Gender"].fillna("Female", inplace = True)
df_with_missing_values_columns

/usr/local/lib/python3.8/dist-packages/pandas/core/generic.py:6392: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return self._update_inplace(result)


Creator Name Creator Gender Language of the Video  \
0              srihari ck           Male               Kannada   
1             Chai Bisket           Male                Telugu   
2             Virda Warsi         Female                 Hindi   
3                     TRT         Female               English   
4                   Vogue         Female               English   
..                    ...            ...                   ...   
900  Sahil Online TV News         Female         Urdu + Arabic   
901  Sahil Online TV News         Female             Nawayathi   
902  Sahil Online Bhatkal         Female             Nawayathi   
903  Sahil Online Bhatkal         Female                  Urdu   
904   Sony Sports Network         Female               English   

    Date of the Last Comment Subtitle  No of Likes  
0                   8/6/2022       No       6700.0  
1                  6/15/2022       No       7800.0  
2                        NaN       No        858.0  
3                  9/22/2022       No        188.0  
4                  9/23/2022      yes     901000.0  
..                       ...      ...          ...  
900                10/4/2022       No        235.0  
901                10/4/2022       No        117.0  
902                 8/4/2022       No          1.0  
903                 8/4/2022       No          1.0  
904                10/4/2022      Yes       7600.0  

[905 rows x 6 columns]

In [11]:
df_with_missing_values_columns.isna().sum()

Creator Name                11
Creator Gender               0
Language of the Video       21
Date of the Last Comment    24
Subtitle                     1
No of Likes                  1
dtype: int64

In [12]:
df_with_missing_values_columns

Creator Name Creator Gender Language of the Video  \
0              srihari ck           Male               Kannada   
1             Chai Bisket           Male                Telugu   
2             Virda Warsi         Female                 Hindi   
3                     TRT         Female               English   
4                   Vogue         Female               English   
..                    ...            ...                   ...   
900  Sahil Online TV News         Female         Urdu + Arabic   
901  Sahil Online TV News         Female             Nawayathi   
902  Sahil Online Bhatkal         Female             Nawayathi   
903  Sahil Online Bhatkal         Female                  Urdu   
904   Sony Sports Network         Female               English   

    Date of the Last Comment Subtitle  No of Likes  
0                   8/6/2022       No       6700.0  
1                  6/15/2022       No       7800.0  
2                        NaN       No        858.0  
3                  9/22/2022       No        188.0  
4                  9/23/2022      yes     901000.0  
..                       ...      ...          ...  
900                10/4/2022       No        235.0  
901                10/4/2022       No        117.0  
902                 8/4/2022       No          1.0  
903                 8/4/2022       No          1.0  
904                10/4/2022      Yes       7600.0  

[905 rows x 6 columns]

The ffill() method replaces the NULL values with the value from the previous row (or previous column, if the axis parameter is set to 'columns').Similarly bfill, backfill and pad methods can also be used. 

In [13]:
df_with_missing_values_columns["Language of the Video"].fillna(method ='ffill', inplace = True)
df_with_missing_values_columns

/usr/local/lib/python3.8/dist-packages/pandas/core/generic.py:6392: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return self._update_inplace(result)


Creator Name Creator Gender Language of the Video  \
0              srihari ck           Male               Kannada   
1             Chai Bisket           Male                Telugu   
2             Virda Warsi         Female                 Hindi   
3                     TRT         Female               English   
4                   Vogue         Female               English   
..                    ...            ...                   ...   
900  Sahil Online TV News         Female         Urdu + Arabic   
901  Sahil Online TV News         Female             Nawayathi   
902  Sahil Online Bhatkal         Female             Nawayathi   
903  Sahil Online Bhatkal         Female                  Urdu   
904   Sony Sports Network         Female               English   

    Date of the Last Comment Subtitle  No of Likes  
0                   8/6/2022       No       6700.0  
1                  6/15/2022       No       7800.0  
2                        NaN       No        858.0  
3                  9/22/2022       No        188.0  
4                  9/23/2022      yes     901000.0  
..                       ...      ...          ...  
900                10/4/2022       No        235.0  
901                10/4/2022       No        117.0  
902                 8/4/2022       No          1.0  
903                 8/4/2022       No          1.0  
904                10/4/2022      Yes       7600.0  

[905 rows x 6 columns]

In [14]:
df3 = df_with_missing_values_columns.fillna(method='ffill')
df3

Creator Name Creator Gender Language of the Video  \
0              srihari ck           Male               Kannada   
1             Chai Bisket           Male                Telugu   
2             Virda Warsi         Female                 Hindi   
3                     TRT         Female               English   
4                   Vogue         Female               English   
..                    ...            ...                   ...   
900  Sahil Online TV News         Female         Urdu + Arabic   
901  Sahil Online TV News         Female             Nawayathi   
902  Sahil Online Bhatkal         Female             Nawayathi   
903  Sahil Online Bhatkal         Female                  Urdu   
904   Sony Sports Network         Female               English   

    Date of the Last Comment Subtitle  No of Likes  
0                   8/6/2022       No       6700.0  
1                  6/15/2022       No       7800.0  
2                  6/15/2022       No        858.0  
3                  9/22/2022       No        188.0  
4                  9/23/2022      yes     901000.0  
..                       ...      ...          ...  
900                10/4/2022       No        235.0  
901                10/4/2022       No        117.0  
902                 8/4/2022       No          1.0  
903                 8/4/2022       No          1.0  
904                10/4/2022      Yes       7600.0  

[905 rows x 6 columns]

In [15]:
df3.isna().sum()

Creator Name                0
Creator Gender              0
Language of the Video       0
Date of the Last Comment    0
Subtitle                    0
No of Likes                 0
dtype: int64

#3) Drop the columns which are irrelevant features to feed into our model.

In [16]:
df.columns

Index(['Video Link', 'Video Views', 'Video Title', 'Channel URL',
       'Creator Name', 'Creator Gender', 'Total Channel Subcribers',
       'Total Chanel Views', 'Duration of Video', 'Duration in Seconds',
       'Date of Video Upload', 'No of Likes', 'Language of the Video',
       'Subtitle', 'Video Description', 'Hashtags', 'No of Comments',
       'Date of the Last Comment', 'Maximum Quality of the Video',
       'No of Videos the Channel', 'No of Playlist', 'Premiered or Not',
       'Community Engagement (Posts per week)',
       'Intern Who Collected the Data'],
      dtype='object')

In [17]:
df.head()

Video Link Video Views  \
0        https://www.youtube.com/watch?v=_0fUOJ55gis     146,059   
1        https://www.youtube.com/watch?v=_70MSLPRfZI     231,969   
2  https://www.youtube.com/watch?v=_8gfCdt3H00&t=87s         878   
3        https://www.youtube.com/watch?v=_CkFNJgKsFs       9,965   
4        https://www.youtube.com/watch?v=_GFkHA5EZdE  39,780,561   

                                         Video Title  \
0    Samse - Didupe Off-road Royal Enfield Himalayan   
1                           Bhoom Baddhal Web Series   
2  Customer Psychology, Four Views Of Consumer De...   
3     Top countries with proven natural gas reserves   
4                     73 Questions With Selena Gomez   

                                         Channel URL Creator Name  \
0                https://www.youtube.com/c/sriharick   srihari ck   
1         https://www.youtube.com/c/ChaibisketTelugu  Chai Bisket   
2  https://www.youtube.com/channel/UCcIUnCsBfONvb...  Virda Warsi   
3                 https://www.youtube.com/c/trtworld          TRT   
4                                    Vogue - YouTube        Vogue   

  Creator Gender Total Channel Subcribers Total Chanel Views  \
0           Male                    40200            7964284   
1           Male                   920000          137535681   
2         Female                      458              72945   
3            NaN                  1680000          541653219   
4            NaN                 12100000         3600496171   

  Duration of Video Duration in Seconds  ... Video Description  Hashtags  \
0         00:30:41             1,841.00  ...               Yes         0   
1         00:14:24               864.00  ...               Yes         1   
2           0:08:12              492.00  ...               Yes         0   
3           0:01:06               66.00  ...               Yes         0   
4           0:07:40              460.00  ...               Yes         0   

  No of Comments Date of the Last Comment Maximum Quality of the Video  \
0            419                 8/6/2022                         1080   
1            189                6/15/2022                         1080   
2              0                      NaN                          720   
3             77                9/22/2022                         1080   
4          28221                9/23/2022                         1080   

   No of Videos the Channel No of Playlist Premiered or Not  \
0                       462           12.0               No   
1                       251           12.0              Yes   
2                       158            0.0               No   
3                     46509           48.0               No   
4                      2872          113.0               No   

   Community Engagement (Posts per week)  Intern Who Collected the Data  
0                                      1                     Akhil Teja  
1                                      2                     Akhil Teja  
2                                      0                    Syed safdar  
3                                     20                    Syed safdar  
4                                      1                    Rahul Singh  

[5 rows x 24 columns]

From above set of columns below columns seems irrelevant to feed into our model:<br>
'Intern Who Collected the Data', 'Video Link','channel URL','duration of video'

In [18]:
df.drop(['Intern Who Collected the Data','Video Link','Channel URL','Duration of Video'], axis=1,inplace=True)
df

Video Views                                        Video Title  \
0       146,059    Samse - Didupe Off-road Royal Enfield Himalayan   
1       231,969                           Bhoom Baddhal Web Series   
2           878  Customer Psychology, Four Views Of Consumer De...   
3         9,965     Top countries with proven natural gas reserves   
4    39,780,561                     73 Questions With Selena Gomez   
..          ...                                                ...   
900       6,892  Calming mashup of Islamic Nasheed 2022 | Isami...   
901      16,554                           Amchi Nawayat Mehfil hee   
902      32,798                                      Aas ge Kaagaz   
903       1,503  Anjuman Tarana by ABHS Students | Anjuman Anju...   
904     429,966  Jhulan Goswami Final Over in ODIs | ENG vs IND...   

             Creator Name Creator Gender Total Channel Subcribers  \
0              srihari ck           Male                    40200   
1             Chai Bisket           Male                   920000   
2             Virda Warsi         Female                      458   
3                     TRT            NaN                  1680000   
4                   Vogue            NaN                 12100000   
..                    ...            ...                      ...   
900  Sahil Online TV News            NaN                   594000   
901  Sahil Online TV News            NaN                   594000   
902  Sahil Online Bhatkal            NaN                    23500   
903  Sahil Online Bhatkal            NaN                    23500   
904   Sony Sports Network            NaN                  3980000   

    Total Chanel Views Duration in Seconds Date of Video Upload  No of Likes  \
0              7964284            1,841.00            7/29/2022       6700.0   
1            137535681              864.00            5/22/2022       7800.0   
2                72945              492.00             2/9/2021        858.0   
3            541653219               66.00            9/13/2022        188.0   
4           3600496171              460.00            3/21/2017     901000.0   
..                 ...                 ...                  ...          ...   
900          129926690              827.00            9/29/2022        235.0   
901          129926690              404.00             8/6/2017        117.0   
902            5104872              384.00             1/3/2021          1.0   
903            5104872              279.00             3/6/2022          1.0   
904          728858923               81.00            9/25/2022       7600.0   

    Language of the Video Subtitle Video Description  Hashtags No of Comments  \
0                 Kannada       No               Yes         0            419   
1                  Telugu       No               Yes         1            189   
2                   Hindi       No               Yes         0              0   
3                 English       No               Yes         0             77   
4                 English      yes               Yes         0          28221   
..                    ...      ...               ...       ...            ...   
900         Urdu + Arabic       No               Yes         0             15   
901             Nawayathi       No               Yes         2              0   
902             Nawayathi       No               Yes         3              0   
903                  Urdu       No               Yes         3              0   
904               English      Yes               Yes         3              0   

    Date of the Last Comment  Maximum Quality of the Video  \
0                   8/6/2022                          1080   
1                  6/15/2022                          1080   
2                        NaN                           720   
3                  9/22/2022                          1080   
4                  9/23/2022                          1080   
..                       ...

#4) Some columns are not statistical data but in the form of strings. It Is difficult for our model to learn when the data is in form of strings. Your task is to encode these strings using proper encoding techniques.

In [19]:
df3.head()

Creator Name Creator Gender Language of the Video Date of the Last Comment  \
0   srihari ck           Male               Kannada                 8/6/2022   
1  Chai Bisket           Male                Telugu                6/15/2022   
2  Virda Warsi         Female                 Hindi                6/15/2022   
3          TRT         Female               English                9/22/2022   
4        Vogue         Female               English                9/23/2022   

  Subtitle  No of Likes  
0       No       6700.0  
1       No       7800.0  
2       No        858.0  
3       No        188.0  
4      yes     901000.0

In [24]:
df3["Creator Gender"] = df3["Creator Gender"].astype('category')
df3["Creator_Gender_encoded"] = df3["Creator Gender"].cat.codes
df3

Creator Name Creator Gender Language of the Video  \
0              srihari ck           Male               Kannada   
1             Chai Bisket           Male                Telugu   
2             Virda Warsi         Female                 Hindi   
3                     TRT         Female               English   
4                   Vogue         Female               English   
..                    ...            ...                   ...   
900  Sahil Online TV News         Female         Urdu + Arabic   
901  Sahil Online TV News         Female             Nawayathi   
902  Sahil Online Bhatkal         Female             Nawayathi   
903  Sahil Online Bhatkal         Female                  Urdu   
904   Sony Sports Network         Female               English   

    Date of the Last Comment Subtitle  No of Likes  Creator_Gender_encoded  
0                   8/6/2022       No       6700.0                       2  
1                  6/15/2022       No       7800.0                       2  
2                  6/15/2022       No        858.0                       1  
3                  9/22/2022       No        188.0                       1  
4                  9/23/2022      yes     901000.0                       1  
..                       ...      ...          ...                     ...  
900                10/4/2022       No        235.0                       1  
901                10/4/2022       No        117.0                       1  
902                 8/4/2022       No          1.0                       1  
903                 8/4/2022       No          1.0                       1  
904                10/4/2022      Yes       7600.0                       1  

[905 rows x 7 columns]

One hot encoding 

In [23]:

one_hot_encoded_data = pd.get_dummies(df3, columns = ['Creator Gender'])
one_hot_encoded_data

Creator Name Language of the Video Date of the Last Comment  \
0              srihari ck               Kannada                 8/6/2022   
1             Chai Bisket                Telugu                6/15/2022   
2             Virda Warsi                 Hindi                6/15/2022   
3                     TRT               English                9/22/2022   
4                   Vogue               English                9/23/2022   
..                    ...                   ...                      ...   
900  Sahil Online TV News         Urdu + Arabic                10/4/2022   
901  Sahil Online TV News             Nawayathi                10/4/2022   
902  Sahil Online Bhatkal             Nawayathi                 8/4/2022   
903  Sahil Online Bhatkal                  Urdu                 8/4/2022   
904   Sony Sports Network               English                10/4/2022   

    Subtitle  No of Likes  Creator_Gender_encoded  Creator Gender_Company  \
0         No       6700.0                       2                       0   
1         No       7800.0                       2                       0   
2         No        858.0                       1                       0   
3         No        188.0                       1                       0   
4        yes     901000.0                       1                       0   
..       ...          ...                     ...                     ...   
900       No        235.0                       1                       0   
901       No        117.0                       1                       0   
902       No          1.0                       1                       0   
903       No          1.0                       1                       0   
904      Yes       7600.0                       1                       0   

     Creator Gender_Female  Creator Gender_Male  
0                        0                    1  
1                        0                    1  
2                        1                    0  
3                        1                    0  
4                        1                    0  
..                     ...                  ...  
900                      1                    0  
901                      1                    0  
902                      1                    0  
903                      1                    0  
904                      1                    0  

[905 rows x 9 columns]